### PR TITLE
UI: Add various preview improvements

### PR DESCRIPTION
**Pull request by VodBox**
### Description

Adds the following changes to the preview:
- Double clicking a source will open it's properties.
- Double clicking a group will open up that group for editing in the preview.
- Optional rotation handle can be enabled in settings.
- Handles and lines will now scale depending on the DPI OBS is running at.

### Motivation and Context

Double clicking a source to open properties saves needing to find the corresponding source in the sources list, and can be quicker to find the source you mean to change properties for.

Editing groups are currently very finicky at best, since there's no way to click into them in the preview.

Rotation handles, while unnecessary for most setups can be useful when needing to change rotations often, and not wanting to use the transform dialog, so is added as an optional handle easily enabled in settings.

Handles and lines are too small on high DPI monitors because they use their standard DPI pixel widths.

### How Has This Been Tested?

This has been tested running Windows 10 64-bit, with many scene collections, old and new.

### Types of changes

 - New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

File: UI/forms/OBSBasicSettings.ui [hunk 1]
```diff
@@ -594,6 +594,13 @@
                       <string>Basic.Settings.General.OverflowSelectionHidden</string>
                      </property>
                     </widget>
+                   </item>
+				   <item row="3" column="1">
```
Comment (hg-pyun) [inline UI/forms/OBSBasicSettings.ui]: This line need to re-indentation.

Comment (VodBox) [reply UI/forms/OBSBasicSettings.ui]: This has been fixed in the latest push.

**Comment by RytoEX** [comment]
If I'm not mistaken, this might address [Mantis 541](https://obsproject.com/mantis/view.php?id=541).

**Comment by jp9000** [comment]
When editing a group, you're unable to switch scenes.  Also I'm not sure how I feel about group editing in general.  Have people asked for this sort of thing?

**Comment by VodBox** [comment]
@jp9000 I'll go ahead and fix the switch scenes things now.

As far as group editing, there's currently no easy way to select items in a group to edit them, since you need to select the item in the sources list. Clicking out in the preview, or clicking the source without dragging will deselect it and you need to go back to the sources list. It's really not a fun time to make changes.

In terms of a dedicated display for editing groups, it's partly my experience with After Effects in terms of what I find natural (double clicking on things to jump into precomps or edit that thing individually), but also just the idea that groups are still scenes and would be nice to edit as such, combined with making things still editable even if an alt crop has been applied.

**Comment by VodBox** [comment]
Scene issues are now fixed.

**Comment by Warchamp7** [comment]
I don't know if I agree with double clicking the preview performing an action. Currently you can click multiple times to cycle through sources

I think a better solution is to scroll the Sources list to a selected sceneitem when it's clicked in the preview

**Comment by VodBox** [comment]
@Warchamp7 

> I don't know if I agree with double clicking the preview performing an action. Currently you can click multiple times to cycle through sources

Double clicking the visual element of something to pull up closer details or to "open it up" is a very common design pattern used in a lot of programs. Most programs don't have cycling, and even in OBS it can still be annoying to get the source you actually want selected in a lot of cases through that method. I'd rather this as a more natural interaction take priority over something non-standard that gets avoided by people selecting with the sources list anyway.

> I think a better solution is to scroll the Sources list to a selected sceneitem when it's clicked in the preview

This really only works as a solution if you have enough sources for it to scroll. Clicking on a source will already make the source in the list active. Even then, it doesn't actually solve the problem I'm trying to solve, which is to have a natural way to open up a source without having to search through a list or jump between parts of the interface.

**Comment by WizardCM** [comment]
Alright, I tested this a while ago and found a few issues. I've just pulled the latest to verify the quirks I ran into.

1. Group edit mode can get confusing if the group is canvas size, or if (like in my case) the group contains white text sources. It makes it super unclear that you're in group edit mode unless you know about it. In general it just feels unintuitive
2. I ran into one issue with Group edit mode that I cannot reproduce, where I could see both the actual group and the "active group" but both scaled at different sizes.
3. The rotation handle currently cannot be grabbed for some reason.

I am a huge fan of both double-click to open properties, and the rotate handle (though I think it should be on by default).

**Comment by PatTheMav** [comment]
Just tested this rebased on current master and looks fine so far, except for

> Double clicking a group will open up that group for editing in the preview

which I haven't fully understood what it's supposed to do or how it works.

**Comment by Warchamp7** [comment]
Have not had the chance to test this yet but some upfront concerns:

- Double clicking a source will open it's properties.

I think this PR is originally from before some other preview improvements. How does double clicking a source interact with clicking multiple times in the preview cycling through all sources the cursor is over


- Double clicking a group will open up that group for editing in the preview.

The fade is a good touch but I don't think it does a good enough job of conveying the functionality. I think this should be broken out into a different PR with some more love to the UX for that kind of usage

- Optional rotation handle can be enabled in settings.
- Handles and lines will now scale depending on the DPI OBS is running at.

These sound good to me

File: UI/window-basic-preview.cpp [hunk 2]
```diff
@@ -1455,13 +1530,71 @@ void OBSBasicPreview::StretchItem(const vec2 &pos)
...
 void OBSBasicPreview::mouseMoveEvent(QMouseEvent *event)
 {
+	OBSBasic *main = reinterpret_cast<OBSBasic *>(App()->GetMainWindow());
 	changed = true;
 
+#ifdef SUPPORTS_FRACTIONAL_SCALING
+	float pixelRatio = main->devicePixelRatioF();
+#else
+	float pixelRatio = main->devicePixelRatio();
+#endif
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp:1591]: This can be simplified, per:
* #4394

File: UI/window-basic-preview.cpp [hunk 3]
```diff
@@ -1558,8 +1692,39 @@ void OBSBasicPreview::leaveEvent(QEvent *)
...
+
 static void DrawSquareAtPos(float x, float y)
 {
+	OBSBasic *main = OBSBasic::Get();
+
+#ifdef SUPPORTS_FRACTIONAL_SCALING
+	float pixelRatio = main->devicePixelRatioF();
+#else
+	float pixelRatio = main->devicePixelRatio();
+#endif
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp:1726]: This can be simplified, per:
* #4394

File: UI/window-basic-preview.cpp [hunk 4]
```diff
@@ -1571,31 +1736,62 @@ static void DrawSquareAtPos(float x, float y)
...
 {
-	float ySide = (y1 == y2) ? (y1 < 0.5f ? 1.0f : -1.0f) : 0.0f;
-	float xSide = (x1 == x2) ? (x1 < 0.5f ? 1.0f : -1.0f) : 0.0f;
+	OBSBasic *main = OBSBasic::Get();
+
+#ifdef SUPPORTS_FRACTIONAL_SCALING
+	float pixelRatio = main->devicePixelRatioF();
+#else
+	float pixelRatio = main->devicePixelRatio();
+#endif
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp:1756]: This can be simplified, per:
* #4394

File: UI/window-basic-preview.cpp [hunk 5]
```diff
@@ -1729,19 +1925,32 @@ bool OBSBasicPreview::DrawSelectedItem(obs_scene_t *scene,
...
 	}
 
-	OBSBasicPreview *prev = reinterpret_cast<OBSBasicPreview *>(param);
 	OBSBasic *main = OBSBasic::Get();
 
+#ifdef SUPPORTS_FRACTIONAL_SCALING
+	float pixelRatio = main->devicePixelRatioF();
+#else
+	float pixelRatio = main->devicePixelRatio();
+#endif
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp:1952]: This can be simplified, per:
* #4394

File: UI/window-basic-preview.cpp [hunk 6]
```diff
@@ -1858,6 +2092,13 @@ bool OBSBasicPreview::DrawSelectedItem(obs_scene_t *scene,
 bool OBSBasicPreview::DrawSelectionBox(float x1, float y1, float x2, float y2,
 				       gs_vertbuffer_t *rectFill)
 {
+	OBSBasic *main = OBSBasic::Get();
+
+#ifdef SUPPORTS_FRACTIONAL_SCALING
+	float pixelRatio = main->devicePixelRatioF();
+#else
+	float pixelRatio = main->devicePixelRatio();
+#endif
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp:2101]: This can be simplified, per:
* #4394

**Review by RytoEX — CHANGES_REQUESTED** [review]
Noticed some places that can be simplified due to dropping support for older Qt versions.

Also, you may need to check for Qt5/6 compatibility in a few places. See:
* #6770

**Comment by cg2121** [comment]
Since we have undo / redo now, I don't think we need the option in the settings for the rotation handle anymore.

**Comment by RytoEX** [comment]
> Since we have undo / redo now, I don't think we need the option in the settings for the rotation handle anymore.

I don't see the connection between undo/redo and whether or not this feature has a setting to enable/disable it.

**Comment by cg2121** [comment]
@jp9000 wanted the rotation handle to be an option, turned off by default, because people might rotate the item accidently and not know how to get it back. With the undo / redo this is no longer a concern.

**Comment by RytoEX** [comment]
> @jp9000 wanted the rotation handle to be an option, turned off by default, because people might rotate the item accidently and not know how to get it back. With the undo / redo this is no longer a concern.

Ah, I see. That said, I'm not sure if I'm going to block on this.

**Comment by SuslikV** [comment]
For the _Preview Scaling_ = _Scale to Window_, the rotation handle rendered outside the 10px border zone, thus it is cropped/almost invisible/inaccessible for full screen sources, or for any sources at the top.

Mentioned "border zone" size is hardcoded to the 10px in OBS:
https://github.com/obsproject/obs-studio/blob/8ef134be7955d3436282c85ff7e806bab2e6f28b/UI/window-basic-main.hpp#L74

**Pull request merged**